### ISSUE TITLE
Overflow behavior and encoding of Double

### ISSUE THREAD

**Issue by drvirgilio** [dhall-lang/dhall-lang#252]
#### Problem

The CBOR encoding of the dhall type `Double` is specified as a [Decimal Fraction](https://tools.ietf.org/html/rfc7049#section-2.4.3). This is an encoding in the form `±m*10^e` where `m` is a bignum, and `e` is a [type 0 or type 1 integer](https://tools.ietf.org/html/rfc7049#section-2.1). The encoding of `e` puts a hard (but *very* large) limit on the size of the number. The range of `e` is [-2^64, 2^64-1]. So what should dhall do when someone inputs a value outside of that range? Here are a few options for resolving this edge case:

#### Options

1.  Silently overflow the exponent and wrap around to negative a negative exponent. This is somewhat unexpected and is how the current implementation operates. See [issue 654](https://github.com/dhall-lang/dhall-haskell/issues/654) in the dhall-haskell repository.
2.  Make this hard limit explicit in the specification and raise an error to users when they input a value which is out of bounds. This seems very limiting if we ever decide we want arithmetic on `Double` since we'd still have to deal with overflow. Also I imagine there's a huge performance penalty compared to floating point operations.
3.  Change the CBOR encoding so that we get unlimited precision. This is a fine response. As an aside in all 3 of the options so far, I'd argue the type should be renamed to `Decimal` or similar so it is clear it is not related to IEEE 754 floating points.
4.  Abandon the `Decimal Fraction` encoding and use the [Double-precision floating-point format](https://en.wikipedia.org/wiki/Double-precision_floating-point_format). This is [supported by CBOR](https://tools.ietf.org/html/rfc7049#section-2.3).
5.  Implement both `Double` and `Decimal` as two available types. An example of how to implement `Decimal` on top of existing dhall types would be a record of two `Integer` such as `{ exponent : Integer, significand : Integer }`.

I'm partial to option 4 or 5 since I expect many people will be immediately converting to a `Double-precision float` in their programs anyway. Doing the needed checks in dhall may ease safety concerns. Also if we ever want dhall to support arithmetic on `Double`, this encoding will become necessary for good performance. So here is a list of a few steps which may be needed for options 4 or 5

#### Steps to standardize on IEEE 754 encoding

1.  A dhall interpreter should be required to raise an error when a user inputs a number which would overflow. E.g `1.8e308` would raise an error immediately when dhall interprets the value. I've seen JSON interpreters do this so it shouldn't be too surprising.
2.  A function from `Text` to `Double` needs to be defined which interprets decimal numbers.
3.  A function from `Double` to `Text` needs to be defined such that there is zero loss of precision. i.e. `text_to_double (double_to_text Q) == Q` for all `Q : Double`, excluding `NaN` and `±Infinity`.
5.  Decide what to do about `NaN` and `±Infinity`. I think it should be disallowed, just like JSON. If someone needs that as a feature they can use a union such as `< number : Double | NaN : {} | negInf : {} | posInf : {} >` or `Optional Double`. Such a wrapper would be needed for [arithmetic](https://en.wikipedia.org/wiki/NaN#Operations_generating_NaN).

Problem 1 should be trivial. For example, in haskell, `read "1.8e308" :: Double` returns `Infinity` which can be used to raise an error. Problems 2-3 have been solved in other contexts. See [this](https://stackoverflow.com/a/7159280) stack overflow answer which lists plenty of resources. Ideally, we would choose algorithms which have widespread adoption. Does anyone know what haskell uses in `show` and `read`, and how it relates to other languages? If not, I'll take a look at the sources in a few places to see what I can find.

#### Conclusion

Wow this post ended up a bit longer than expected. Anyone have any thoughts or opinions?

**Comment by Gabriel439** [comment]
This is a tough decision! 🙂 

My (weak) preference is for option 4.  In other words, we change `Double` and have it mean an IEEE754 floating point and we drop support for unlimited precision decimal fractions.  We can later reintroduce support for unlimited precision decimal fractions for them upon user request if users need to do so.  Also, like you mentioned, users can always embed unlimited precision decimal fractions within the language using `{ exponent : Integer, significand : Integer }`.

There is only one minor issue that I can think of: if we parse decimal literals into, say, Haskell `Double`s and then re-render them as `Text` in Dhall using `Double/show` the representation will change trip due to loss of precision.  Technically this is not a violation of the current standard because the standard actually only guarantees that the value will round trip in the other direction.  In other words, the standard only guarantees that if you render a `Double` value using `Double/show` then the generated `Text` literal is valid source code for the underlying `Double` value.

> A dhall interpreter should be required to raise an error when a user inputs a number which would overflow. E.g `1.8e308` would raise an error immediately when dhall interprets the value. I've seen JSON interpreters do this so it shouldn't be too surprising.

I agree.  I think it's sensible to disallow `Double` literals that are outside the valid floating point range.  That's a useful safety guarantee.

> A function from `Text` to `Double` needs to be defined which interprets decimal numbers.

We should probably not provide a function to convert `Text` values into `Double`s (i.e. no `Double/read : Text -> Optional Double`), mainly because I try to avoid providing any way to parse or otherwise introspect `Text` values to avoid stringly-typed programming.

> A function from `Double` to `Text` needs to be defined such that there is zero loss of precision. i.e. `text_to_double (double_to_text Q) == Q` for all `Q : Double`, excluding `NaN` and `±Infinity`.

This function exists and is `Double/show`

> Decide what to do about `NaN` and `±Infinity`. I think it should be disallowed, just like JSON. If someone needs that as a feature they can use a union such as `< number : Double | NaN : {} | negInf : {} | posInf : {} >` or `Optional Double`. Such a wrapper would be needed for arithmetic.

I actually don't mind supporting `NaN` or `±Infinity`.  Note that even if users are incapable of creating those values entirely within `Dhall`, a language binding will probably want to be able to convert values from the Host language into Dhall (i.e. the `Inject` class in the Dhall API, analogous to `ToJSON`).  This is why I think that Dhall should be capable of handling the full range of `Double` values.

Also, if Dhall ever does support arithmetic it will likely not be for `Double` values, but rather some unlimited precision alternative such as decimal fractions so that the language doesn't have to deal with overflow.  The main reason that the language supports `Double` at all is purely so that Dhall configuration files can store them, but otherwise the host language is responsible for dealing with the messiness of floating point arithmetic.

**Comment by drvirgilio** [comment]
I think I was unclear and confusing myself on steps 2-3. I meant functions in an implementation of dhall which is similar to haskell's `show` and `read` for type `Double`. (ignoring that `read` doesn't allow prepended `+`). Thinking about it it's really a non issue. Implementers just need to convert the string to closest `Double`. And for output they can use `show`, which is already defined in the standard such that `read . show = id`. But an exception needs to be added for `NaN` which I explain in point 2 below.

And to solve the minor issue you mentioned, a feature request for `dhall lint` might be to display a warning when a user inputs a `Double` which loses precision. For example, `1.0000000000000005`, once converted to `Double` and back to `String` would show as `1.0000000000000004` so `dhall lint` produces a warning. That's more complicated of a request, (I think it requires normalization of the decimal string to compare it to the output of `show`). But it would be another way to push new users into learning about floating point imprecision :)

And I see two potential problems with supporting `NaN` or `±Infinitiy`.

1.  How would dhall-to-json support work? Would it just output `null` in all 3 cases? That loses information but might be more ergonomic than wrapping every double up in another object which fully encodes them. I've seen other encoders just use `null`, so this would be a fine choice.
2.  `NaN` is encoded as a double where the exponent bits are all 1 and the significand (mantissa) is any non-zero number, so `NaN` has 2^52-1 different encodings (crazy right!). This is probably a non-issue in the vast majority of cases, so I expect it should be normalized to a single encoding for the purpose of a semantic hash. If someone in the future really needs to discern between different NaN values, we can worry about it then. E.g. a CBOR to CBOR dhall interpreter could pass the same `NaN` value along without modification.

#### Conclusion

How does this sound for an action plan for adjustment to the standard and implementation?

1.  Require an immediate error when interpreting a `Double` larger or smaller than the min or max values.
2.  Add the grammar for interpreting the exceptions. This is an aesthetic choice. `NaN`, `-Infinity`, `+Infinity`. Or perhaps shortened to `+Inf`, `-Inf`. I'd say require the `+` so we use fewer reserved words.
3.  Require that `NaN` normalizes to single, specific bit string for the purpose of semantic hashing.
4.  Specify the exception that `read (show (Q :: Double))` won't necessarily re-encode to the same bits in the case of `NaN`
5.  Allow the use of `null` when outputting a `Double` in dhall-to-json
6.  Add a warning in `dhall lint` in the case that an input `Double` would lose a significant amount of precision.

**Comment by Gabriel439** [comment]
@drvirgilio: Yeah, those seem reasonable to me.  I would only propose three minor changes:

First, I checked to see what `jq` does for those three values, and I got:

```
$ jq . <<< "NaN"
null

$ jq . <<< "Infinity"
1.7976931348623157e+308

$ jq . <<< "-Infinity"
-1.7976931348623157e+308
```

... which seems like reasonable behavior to me that `dhall-to-json` could mirror.

Second, I think `dhall lint` should automatically fix issues rather than warn about them, so if we treat a loss of precision as a warning then we should fix it by automatically correcting the `Double` value.

Third, rather than requiring that `read (show (Q :: Double)) = Q` and adding an exception for `NaN`, we can instead require that:

```
show (read (show (X :: Double))) = show X

read (show (read (Y :: Text))) = read Y
```

**Comment by drvirgilio** [comment]
:+1: Those minor changes sound good to me. Matching the behavior of `jq` seems like a reasonable decision. I've forked the standard to write changes. I'll issue a pull request for review when I have something.

**Comment by drvirgilio** [comment]
I took a stab at both the `dhall` implementation and the standard. Let me know what you think. I haven't done anything with `dhall-json` or the like so that still needs doing. I figured I show you what I have before doing that. I'll address each step from above for both implementation and the standard:

1.  Implemented bounds check but didn't change the standard yet
2.  Done in both. Added grammar for `NaN` and `[-]Infinity`. Didn't add grammar for `+Infinity` because it wasn't playing nice with expressions such as `let InfinityVar = 5 in 2+InfinityVar`. It would parse the `+Infinity` as a double instead of a variable like it would in any other case. Fixing that seems non-trivial.
3.  This was actually simple because `cborg` automatically picks a canonical form of `0x7e00` when giving any `NaN` value so I put that in the standard and decided to embrace using the canonical form as described in the CBOR RFC [here](https://tools.ietf.org/html/rfc7049#section-3.9).
4.  Still needs put in the standard but the implementation seems to work fine.
5.  Still needs doing
6.  `dhall format` converts to and from `Double` so it puts the values into a canonical representation and removes any precision that can't be stored in a `Double`
